web/api/v1/api.go

```diff
@@ -464,6 +475,82 @@ func (api *API) queryRange(r *http.Request) (result apiFuncResult) {
 	}, nil, res.Warnings, qry.Close}
 }
 
+func (api *API) queryExemplars(r *http.Request) apiFuncResult {
+	start, err := parseTime(r.FormValue("start"))
+	if err != nil {
+		err = errors.Wrapf(err, "invalid parameter 'start'")
+		return apiFuncResult{nil, &apiError{errorBadData, err}, nil, nil}
+	}
+	end, err := parseTime(r.FormValue("end"))
+	if err != nil {
+		err = errors.Wrapf(err, "invalid parameter 'end'")
+		return apiFuncResult{nil, &apiError{errorBadData, err}, nil, nil}
+	}
+	if end.Before(start) {
+		err := errors.New("end timestamp must not be before start time")
+		return apiFuncResult{nil, &apiError{errorBadData, err}, nil, nil}
+	}
+
+	ctx := r.Context()
+	if to := r.FormValue("timeout"); to != "" {
+		var cancel context.CancelFunc
+		timeout, err := parseDuration(to)
+		if err != nil {
+			err = errors.Wrap(err, "invalid parameter 'timeout'")
+			return apiFuncResult{nil, &apiError{errorBadData, err}, nil, nil}
+		}
+
+		ctx, cancel = context.WithTimeout(ctx, timeout)
+		defer cancel()
+	}
+
+	q, err := api.Queryable.Querier(ctx, timestamp.FromTime(minTime), timestamp.FromTime(maxTime))
+	if err != nil {
+		return apiFuncResult{nil, &apiError{errorBadData, err}, nil, nil}
+	}
+	defer q.Close()
+
+	expr, err := parser.ParseExpr(r.FormValue("query"))
+	if err != nil {
+		return apiFuncResult{nil, &apiError{errorBadData, err}, nil, nil}
+	}
+
+	selectors, err := parser.ExtractSelectors(expr)
+	if err != nil {
+		return apiFuncResult{nil, &apiError{errorBadData, err}, nil, nil}
+	}
+	if len(selectors) < 1 {
+		return apiFuncResult{nil, &apiError{errorBadData, fmt.Errorf("no series found for supplied query: %s", expr)}, nil, nil}
+	}
+
+	var retExemplars []exemplarData
+	var series []storage.SeriesSet
+
+	for _, selectorArr := range selectors {
+		series = append(series, q.Select(false, nil, selectorArr...))
```

Comment:
This is fairly expensive currently, the same as if you were actually fetching the data. https://github.com/prometheus/prometheus/issues/5547#issuecomment-620032812 may need to go in first.

Reply:
I'll keep an eye on that issue, thanks. Unless there's active work happening there, and given that exemplars won't be widely used soon, I'll continue with this as is for now.

Reply:
Looks like I need to add a `hints.Func: "series"` here.